nomt/src/page_cache.rs

```diff
@@ -0,0 +1,251 @@
+///   are notified with the dirty page.
+/// - Completed. The page has been fetched and the waiters are notified with the fetched page.
+struct InflightFetch {
+    tx: Option<crossbeam_channel::Sender<Page>>,
+    rx: crossbeam_channel::Receiver<Page>,
+    supplanted: Option<Page>,
+}
+
+impl InflightFetch {
+    fn new() -> Self {
+        let (tx, rx) = crossbeam_channel::bounded(1);
+        Self {
+            tx: Some(tx),
+            rx,
+            supplanted: None,
+        }
+    }
+
+    /// Called when the page is updated before the fetch is completed.
+    ///
+    /// Can be called once. Panics otherwise.
+    fn supplant_and_notify(&mut self, page: Page) {
+        assert!(self.supplanted.is_none());
+        self.supplanted = Some(page.clone());
+        if let Some(tx) = self.tx.take() {
+            let _ = tx.send(page);
+        }
+    }
+
+    /// Notifies all the waiting parties that the page has been fetched and destroys this handle.
+    fn complete_and_notify(mut self, page: Page) {
+        if let Some(tx) = self.tx.take() {
+            let _ = tx.send(page);
+        }
+    }
+
+    /// Returns the promise that resolves when the page is fetched.
+    fn promise(&self) -> PagePromise {
+        match &self.supplanted {
+            Some(page) => PagePromise::Now(page.clone()),
+            None => PagePromise::Later(self.rx.clone()),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct PageCache {
+    shared: Arc<Mutex<Shared>>,
+    store: Store,
+    /// The thread pool used for fetching pages from the store.
+    ///
+    /// Used for limiting the number of concurrent page fetches.
+    fetch_tp: ThreadPool,
+}
+
+struct Shared {
+    /// The unmodified pages loaded from the store.
+    ///
+    /// This cache must be always coherent with the underlying store. That implies that the updates
+    /// performed to the store must be reflected in the cache.
+    prestine: HashMap<PageId, Page>,
+    /// The pages that were modified, but not yet committed.
+    dirty: HashMap<PageId, Page>,
+    /// Pages that are currently being fetched from the store.
+    inflight: HashMap<PageId, InflightFetch>,
+}
+
+impl PageCache {
+    pub fn new(store: Store, o: &Options) -> Self {
+        let shared = Arc::new(Mutex::new(Shared {
+            prestine: HashMap::new(),
+            dirty: HashMap::new(),
+            inflight: HashMap::new(),
+        }));
+        let fetch_tp = ThreadPool::new(o.fetch_concurrency);
+        Self {
+            shared,
+            store,
+            fetch_tp,
+        }
+    }
+
+    /// Retrieves the page data at the given [`PageId`] asynchronously.
+    ///
+    /// If the page is in the cache, it is returned immediately. If the page is not in the cache, it
+    /// is fetched from the underlying store and returned.
+    fn retrieve(&self, page_id: PageId) -> PagePromise {
+        let mut shared = self.shared.lock().unwrap();
```

Comment:
we could consider a switch to `parking_lot` just to avoid these annoying poisoning unwraps. it is supposedly faster too.

Reply:
good call

Reply:
https://github.com/thrumdev/nomt/pull/40